streaming/java/streaming-state/src/main/java/org/ray/streaming/state/backend/TransactionKeyStateBackend.java

```diff
@@ -0,0 +1,193 @@
+package org.ray.streaming.state.backend;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import org.ray.streaming.state.ITransactionState;
+import org.ray.streaming.state.keystate.desc.AbstractStateDescriptor;
+import org.ray.streaming.state.keystate.desc.AbstractStateDescriptor.DescType;
+import org.ray.streaming.state.keystate.state.facade.ListStateFacade;
+import org.ray.streaming.state.keystate.state.facade.MapStateFacade;
+import org.ray.streaming.state.keystate.state.facade.ValueStateFacade;
+import org.ray.streaming.state.store.IKMapStore;
+import org.ray.streaming.state.store.IKVStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Transaction support primitive operations like finish, commit, ackcommit and rollback.
+ */
+public abstract class TransactionKeyStateBackend implements ITransactionState {
```

Comment:
Is this class `ThreadSafe` or must be executed single-threaded outside? I notice some of the methods(like `commit`, `askCommit`) require iterating internal data structure which is potentially run into race.

Reply:
State VALUE modification is not thread safe. By default, every processing thread has its own
space to handle state. 

I'll add the comment